File: src/child_workflow/child_workflow.erl
-module(child_workflow).

% elp:ignore W0012 W0040 E1599
-moduledoc {file, "../../docs/child_workflow.md"}.

-export([
    run/1
]).

-spec run(ChildWorkflowOutcome :: complete | cancel | fail | invalid | atom()) ->
    temporal_sdk:workflow_result() | no_return().
run(ChildWorkflowOutcome) when is_atom(ChildWorkflowOutcome) ->
    case
        temporal_sdk:start_workflow(
            cluster_1, "default", child_workflow_workflow, [
                wait, {input, [[atom_to_binary(ChildWorkflowOutcome)]]}
            ]
        )
    of
        {S, _} when S =:= error; S =:= ok -> throw("Unexpected result.");
        {_StartWorkflowRet, WorkflowResult} -> WorkflowResult
    end.

File: lib/child_workflow/child_workflow.ex
defmodule ChildWorkflow do
  @external_resource "docs/child_workflow.md"
  @moduledoc File.read!("docs/child_workflow.md")

  @spec run(child_workflow_outcome :: :complete | :cancel | :fail | :invalid | atom()) ::
          :temporal_sdk.workflow_result() | no_return()
  def run(child_workflow_outcome) when is_atom(child_workflow_outcome) do
    r =
      TemporalSdk.start_workflow(:cluster_1, "default", ChildWorkflow.Workflow, [
        :wait,
        input: [[:erlang.atom_to_binary(child_workflow_outcome)]]
      ])

    case r do
      {s, _} when s === :error or S === :ok -> throw("Unexpected result.")
      {_start_workflow_ret, workflow_result} -> workflow_result
    end
  end
end
